### ISSUE TITLE
Point outline width is not sharp

### ISSUE THREAD

**Issue by nvkelso** [tangrams/tangram#577]
Following up on a PR comment https://github.com/tangrams/tangram/pull/508#discussion_r131962936 after it was closed:

From @javisantana:

Hey, sorry to open this conversation but I don't fully understand this.

`v_aa_factor` is calculated taking like `TANGRAM_PX_FADE_RANGE * 1.0/ length(a_shape.xy / 256.)`. 

This opens some questions:
- even `v_aa_factor` is a varying, does it vary? the v_aa_factor calculated for each vertex should be the same (I'm talking just about the quads used to render a point)
- when point size is <= 2px `v_aa_factor` is greater than 1 so the `smoothstep(1. - v_aa_factor, 1., point_dist)` is kind of misleading
- if `a_shape` is [+-128, +-128], then should n't be `TANGRAM_PX_FADE_RANGE * 1.0/ length(a_shape.xy / 128.)` or something like `max(1.0, TANGRAM_PX_FADE_RANGE * 1.0/ length(a_shape.xy / 256.))` to avoid aliasing on 1px points? 

Am I misunderstanding `v_aa_factor` usage?

thanks!

**Comment by nvkelso** [comment]
I'm having the same problem today as I experiment with https://mapzen.com/tangram/view/?api=506/875#7.623/37.463/-122.311.

I'm finding the widths are:

- Too big compared to the pixel sizes specified in the scene file
- Blurry (not pixel perfect crisp)

<img width="259" alt="screen shot 2017-08-12 at 10 50 26" src="https://user-images.githubusercontent.com/853051/29243101-2f9ebbbc-7f4c-11e7-8eca-1206b337f6a2.png">


Here's I'm overcompensating by choosing small values that are still showing up larger than expected:

```
                outline:
                    color: black
                    width: [[4,0.25px],[6,0.25px],[8,0.25px],[10,1px]]
```

**Comment by bcamper** [comment]
The basic answer to both of your questions is that these factors were mostly tuned by hand (by @patriciogonzalezvivo and I) to get a balance between aliasing (too blocky) and antialiasing (too blurry). The fade is currently configured to show a *full width line* of the width specified in `outline.width`, with the antialiasing feathering out from that. They can certainly be further tuned and @javisantana you may be right that limiting the factor on smaller points will help (have you tried? ;) 

However, we do need to keep in mind (relevant to @nvkelso's example) that there is a minimum size (for both the point fill and the line around it) at which we can get satisfactory results. I don't think a line less than `1px`(or `0.5px` on a 2x display, since Tangram uses logical pixels) will ever be really crisp, and will probably just muddle with the point fill color.

Agree we should try tuning for smaller points though.

Further implementation notes @javisantana:

`a_shape.xy` is half the width and height of the sprite, in pixels, multiplied by 256. Specifically, for each vertex it's an amount in pixels to offset from the center of the sprite. The actual position of the vertex (`a_position`) is the sprite's center, so that 3d transforms can first be applied, and then a 2d offset in screen-space can be applied afterwards. The multiplication by 256 is because it is stored as an integer (to save space in the GL vertex buffer), so it is multiplied by some "base" number in the CPU, and then divided by that same number in the shader, to get back to a floating point number (this is a fairly common GL technique and you'll see it around several other parts of Tangram, since most vertex attributes don't need a full 4-byte float; it's also similar to the old-school pre-floating point technique of ["fixed integer" math](https://en.wikipedia.org/wiki/Fixed-point_arithmetic)).

**Comment by javisantana** [comment]
ok, thanks for the update. I tried some stuff but I needed to know a little bit more about the rationale behind it.

I've been comparing with the way mapnik (cairo) does things and there are differenteces. I tried to tweak the values to get the same but it's tricky without knowing the internal details. 

I will send a PR with some proposed fixes and we can discuss things there.

**Comment by bcamper** [comment]
Thanks, happy to see a PR. I think the basic logic was in @donflopez's original PR, but the antialiasing was definitely added/changed after. 

Also, to your question about "does `v_aa_factor` actually vary?": no, it does not! You're correct that it's computed as the same value at each corner of the sprite. But the (perhaps poorly named) `varying` is the only way for a GL vertex shader to pass info to a fragment shader (whether it varies or not). And precisely because this value doesn't change across the sprite, we want to only compute it in the vertex shader to reduce computation. We could also calculate it CPU-side and put it in a vertex attribute, if we had an extra slot for it (I have a separate PR that may actually be freeing one up soon). But it should also be very fast to calculate once per vertex on the GPU (since most scenes have far fewer "points"/sprites than they do polygon + line geometries).

**Comment by bcamper** [comment]
Should be addressed in v0.14.0 with #583

**Comment by bcamper** [comment]
Addressed in [v0.13.3](https://github.com/tangrams/tangram/releases/tag/v0.13.3)
